### DISCUSSION TITLE
How to download images from Strapi server during `nuxi generate`

### DISCUSSION THREAD

**Discussion by Zerro97** [nuxt/nuxt#16371] [Questions]
### Question:
Is there an option in Nuxt3 to fetch/download the image files during build?

### Context:
To give background on the project, I'm using:

1. Nuxt for SSG
2. Strapi for CMS
3. Nginx for serving static contents
4. Ubuntu dedicated server to deploy my project to

After making sure that Strapi Server is running (`strapi start`), I have generated the static contents(`npx nuxi generate`). I then copied `.output/public` folder to the ubuntu server and let nginx serve the static contents. 

As for the source code, I have wrapped my Strapi fetches with useAsyncData to ensure that fetching content happens during build.

### Problem:
While the application has fetched the content from Strapi server during build, I realized that there is still a problem with referring to image files in production. The problem is that, during `nuxi generate`, it only fetches information about the image files (ex. image url) and not downloading the file itself. I was hoping that `nuxi generate` will create image files in a `.output/public` folder, but it wasn't the case. So without running Strapi server in production, I cannot refer to the image files as the files are hosted in Strapi server. 

---
I'm hoping to not fetch any resources during production environment but as of now, I still need to fetch images hosted in Strapi server. Is there a way to download images from Strapi server during `nuxi generate`?

---
### Misc
<details>
<summary></summary>

By the way, looking forward to the stable release tomorrow!
![Screen Shot 2022-11-15 at 12 07 20 PM](https://user-images.githubusercontent.com/34781112/201817814-78b34768-6e4f-406e-9af9-f15432257966.png)
</details>

**Comment by Zerro97** [comment]
I just discovered that [NuxtImage](https://image.nuxtjs.org/providers/strapi) have Strapi support. Maybe using Nuxt Image can resolve my issue but it seems like it doesn't support `nuxi generate` yet. Guess I will wait until [this PR](https://github.com/nuxt/image/pull/614) resolves..

**Comment by MrSunshyne** [comment]
@Zerro97 any luck with this? I've been doing to achieve the following, without success so far: https://github.com/nuxt/nuxt/discussions/16061#discussioncomment-4948099

**Reply by Zerro97** [reply]
Hi! I think the [PR](https://github.com/nuxt/image/pull/614) that I mentioned above was to provide support for pre-rendering/optimizing images that were inside the `public` folder (filesystem), not for individual CMS providers or images that are linked by http link. (so I can't use NuxtImage to prerender images that are fetched from strapi. Only works for images inside public folder)

So before this PR, the images inside `public` folder was not generated into the `.output/public` folder when using `nuxt generate` command. After the PR, they are generated in folder like `├─ /_ipx/_/test.png (132ms)` inside `.output/public`

Also note that you need to use the default provider (IPX) to generate the optimized images in `.output/public` folder (generated from `public` folder). I think this doesn't work with any other provider apart from IPX.

At least that's what I understood by looking at some of the source code in NuxtImage like [here](https://github.com/nuxt/image/blob/main/src/runtime/ipx.ts):

```javascript
const ipxOptions = {
  ...(useRuntimeConfig().ipx || {}),
  dir: fileURLToPath(new URL(process.env.prerender ? '../../.output/public' : '../public', import.meta.url))
}
```

and [here](https://github.com/unjs/ipx/blob/main/src/ipx.ts) (from unjs/IPX):
```javascript
const defaults = {
  dir: getEnvironment("IPX_DIR", "."),
  domains: getEnvironment("IPX_DOMAINS", []),
  alias: getEnvironment("IPX_ALIAS", {}),
  fetchOptions: getEnvironment("IPX_FETCH_OPTIONS", {}),
  maxAge: getEnvironment("IPX_MAX_AGE", 300),
  sharp: {}
};
// ..
// ..
if (options.dir) {
  context.sources.filesystem = createFilesystemSource({
    dir: options.dir,
    maxAge: options.maxAge
  });
}
if (options.domains) {
  context.sources.http = createHTTPSource({
    domains: options.domains,
    fetchOptions: options.fetchOptions,
    maxAge: options.maxAge
  });
}
```
